core/trino-main/src/main/java/io/trino/sql/planner/ArraySortAfterArrayDistinctRewriter.java

```diff
@@ -0,0 +1,99 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.trino.sql.planner;
+
+import com.google.common.collect.ImmutableMap;
+import io.trino.Session;
+import io.trino.metadata.Metadata;
+import io.trino.operator.scalar.ArrayDistinctFunction;
+import io.trino.operator.scalar.ArraySortFunction;
+import io.trino.spi.type.Type;
+import io.trino.sql.tree.Expression;
+import io.trino.sql.tree.ExpressionRewriter;
+import io.trino.sql.tree.ExpressionTreeRewriter;
+import io.trino.sql.tree.FunctionCall;
+import io.trino.sql.tree.NodeRef;
+import io.trino.sql.tree.QualifiedName;
+import io.trino.sql.tree.SymbolReference;
+
+import java.util.List;
+import java.util.Map;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public final class ArraySortAfterArrayDistinctRewriter
+{
+    public static Expression rewrite(Expression expression, Map<NodeRef<Expression>, Type> expressionTypes, Metadata metadata)
+    {
+        return ExpressionTreeRewriter.rewriteWith(new Visitor(expressionTypes, metadata), expression);
+    }
+
+    private ArraySortAfterArrayDistinctRewriter() {}
+
+    public static Expression rewrite(Expression expression, Session session, Metadata metadata, TypeAnalyzer typeAnalyzer, TypeProvider typeProvider)
+    {
+        requireNonNull(metadata, "metadata is null");
+        requireNonNull(typeAnalyzer, "typeAnalyzer is null");
+
+        if (expression instanceof SymbolReference) {
+            return expression;
+        }
+        Map<NodeRef<Expression>, Type> expressionTypes = typeAnalyzer.getTypes(session, typeProvider, expression);
+
+        return rewrite(expression, expressionTypes, metadata);
+    }
+
+    private static class Visitor
+            extends ExpressionRewriter<Void>
+    {
+        private final Map<NodeRef<Expression>, Type> expressionTypes;
+        private final Metadata metadata;
+
+        public Visitor(Map<NodeRef<Expression>, Type> expressionTypes, Metadata metadata)
+        {
+            this.expressionTypes = ImmutableMap.copyOf(requireNonNull(expressionTypes, "expressionTypes is null"));
+            this.metadata = metadata;
+        }
+
+        @Override
+        public Expression rewriteFunctionCall(FunctionCall node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            if (metadata.getFunctionMetadata(metadata.decodeFunction(node.getName())).getActualName().equals(ArrayDistinctFunction.NAME)
+                    && node.getArguments().size() > 0
+                    && node.getArguments().get(0) instanceof FunctionCall
+                    && metadata.getFunctionMetadata(metadata.decodeFunction(((FunctionCall) node.getArguments().get(0)).getName())).getActualName().equals(ArraySortFunction.NAME)) {
```

Comment:
The name "array_sort" identifies two different functions. According to Trino docs, there are:
`array_sort(x) → array`
`array_sort(array(T), function(T, T, int)) -> array(T)`

Make sure that only the first argument goes to `array_distinct`, and the second argument, if present, remains in the `array_sort` call.
Add test for such case.

Reply:
Hi @kasiafi , I've implemented support for this, however i've not managed to get the test working (See the commented out test). The test is faling with java.lang.IllegalStateException: missing expression for alias x. If you would provide any pointers on how to solve that one, that would be helpfull. :)

Reply:
Apparently, the assertions cannot handle a lambda.

I think that you could add this test case in `query`. Also, you could add some more examples there, with non-distinct and unsorted arrays as input, to show that the results are correct.